src/FuelClient.cc

```diff
@@ -687,6 +687,60 @@ Result FuelClient::DownloadModel(const ModelIdentifier &_id,
   if (!this->dataPtr->cache->SaveModel(newId, resp.data, true))
     return Result(ResultType::FETCH_ERROR);
 
+  // Locate any dependencies from the input model and download them.
+  std::string path;
+  ignition::msgs::FuelMetadata meta;
+  if (this->CachedModel(ignition::common::URI(newId.UniqueName()), path))
+  {
+    std::string metadataPath =
+      ignition::common::joinPaths(path, "metadata.pbtxt");
+    std::string modelConfigPath =
+      ignition::common::joinPaths(path, "model.config");
+    if (ignition::common::exists(metadataPath))
+    {
+      // Read the pbtxt file.
+      std::ifstream inputFile(metadataPath);
+      std::string inputStr((std::istreambuf_iterator<char>(inputFile)),
+        std::istreambuf_iterator<char>());
+
+      // Parse the file into the fuel metadata message
+      google::protobuf::TextFormat::ParseFromString(inputStr, &meta);
+
+      for (int i = 0; i < meta.dependencies_size(); i++)
+      {
+        std::string dependencyPath;
+        ignition::common::URI dependencyURI(meta.dependencies(i).uri());
+
+        // If the model is not already cached, download it; this prevents
+        // any sort of cyclic dependencies fromo running infinitely
+        if (!this->CachedModel(dependencyURI, dependencyPath))
+          this->DownloadModel(dependencyURI, dependencyPath);
+      }
+    }
+    else if (ignition::common::exists(modelConfigPath))
+    {
+      std::ifstream inputFile(modelConfigPath);
+      std::string inputStr((std::istreambuf_iterator<char>(inputFile)),
+          std::istreambuf_iterator<char>());
+
+      if (!ignition::msgs::ConvertFuelMetadata(inputStr, meta))
+      {
+        return Result(ResultType::UPLOAD_ERROR);
+      }
+
+      for (int i = 0; i < meta.dependencies_size(); i++)
```

Comment:
This `for` loop is a duplicate of the one above. Would it be possible to move this into a lambda function or re-arrange the logic in order to prevent the duplication?

Reply:
Logic condensed here d8c0dca9933eea97b187373b083db3202e100a56